File: Programming Fundamentals/Training/09. Sets And Dictionaries - Adv C# May 2016/LegendaryFarming.cs
using System;
using System.Collections.Generic;
using System.Linq;

class LegendaryFarming
{
    static void Main()
    {
        var keyMaterials = new Dictionary<string, int>();
        var junk = new SortedDictionary<string, int>();

        CollectMaterials(keyMaterials, junk);
        PrintRemainingMaterials(keyMaterials, junk);
    }

    private static void PrintRemainingMaterials(Dictionary<string, int> keyMaterials, SortedDictionary<string, int> junk)
    {
        var sortedKeMaterials = keyMaterials
            .OrderByDescending(qnty => qnty.Value)
            .ThenBy(material => material.Key);
        foreach (var material in sortedKeMaterials)
        {
            Console.WriteLine($"{material.Key}: {material.Value}");
        }

        foreach (var item in junk)
        {
            Console.WriteLine($"{item.Key}: {item.Value}");
        }

    }

    private static void CollectMaterials(Dictionary<string, int> keyMaterials, SortedDictionary<string, int> junk)
    {
        keyMaterials.Add("shards", 0);
        keyMaterials.Add("fragments", 0);
        keyMaterials.Add("motes", 0);

        while (true)
        {
            var items = Console.ReadLine()
                .ToLower()
                .Split();

            for (int i = 0; i < items.Length; i += 2)
            {
                int quantity = int.Parse(items[i]);
                string material = items[i + 1];

                if (keyMaterials.ContainsKey(material))
                {
                    FillKeyMaterials(keyMaterials, quantity, material);
                }
                else
                {
                    FillJunk(junk, material, quantity);
                }

                if (keyMaterials.ContainsKey(material) && keyMaterials[material] >= 250)
                {
                    string legendaryItem = "";
                    switch (material)
                    {
                        case "shards":
                            legendaryItem = "Shadowmourne"; break;
                        case "fragments":
                            legendaryItem = "Valanyr"; break;
                        case "motes":
                            legendaryItem = "Dragonwrath"; break;
                    }
                    Console.WriteLine($"{legendaryItem} obtained!");
                    keyMaterials[material] -= 250;
                    return;
                }
            }
        }
    }

    private static void FillJunk(SortedDictionary<string, int> junk, string material, int quantity)
    {
        if (junk.ContainsKey(material))
        {
            junk[material] += quantity;
        }
        else
        {
            junk.Add(material, quantity);
        }
    }

    private static void FillKeyMaterials(Dictionary<string, int> keyMaterials, int quantity, string material)
    {
        keyMaterials[material] += quantity;
    }
}

File: Java_Fundamentals_January_2017/Java_Advanced/Exersices/04. Sets and Maps/src/LegendaryFarming.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class LegendaryFarming {
    public static void main(String[] args) throws IOException {
        BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in));

        Map<String, Integer> keyMaterials = new HashMap<>();
        keyMaterials.put("shards", 0);
        keyMaterials.put("fragments", 0);
        keyMaterials.put("motes", 0);


        Map<String, Integer> junkMaterials = new HashMap<>();

        boolean isObtained = false;
        String obtainedItem = null;

        Set<String> keys =
                new HashSet<String>(){{add("shards"); add("fragments"); add("motes");}};
        Map<String, String> items = new HashMap<>();
        items.put("shards", "Shadowmourne");
        items.put("fragments", "Valanyr");
        items.put("motes", "Dragonwrath");


//        BiConsumer<Map<String, Integer>, String> x = (m, s) -> {
//            if (!m.containsKey(material)) {
//                m.put(material, qty);
//            } else {
//                m.put(material, keyMaterials.get(material) + qty);
//            }
//        }

        while(true){

            String[] tokens = in.readLine().toLowerCase().split("\\s+");

            for(int i= 0; i < tokens.length; i += 2) {
                int qty = Integer.parseInt(tokens[i]);
                String material = tokens[i + 1];

                Map<String, Integer> currentContainer
                        = keys.contains(material) ? keyMaterials : junkMaterials;
                if (!currentContainer.containsKey(material)) {
                    currentContainer.put(material, qty);
                } else {
                    currentContainer.put(material, currentContainer.get(material) + qty);
                }

                if(keyMaterials.containsKey(material) && keyMaterials.get(material) >= 250 && !isObtained){
                    isObtained = true;
                    obtainedItem = items.get(material);
                    keyMaterials.put(material, keyMaterials.get(material) - 250);
                    break;
                }
            }

            if(isObtained) break;
        }

        System.out.printf("%s obtained!\n", obtainedItem);

        keyMaterials.entrySet().stream()
                .sorted((m1, m2) -> {
                    if (m1.getValue() > m2.getValue()){
                        return -1;
                    } else if(m1.getValue() < m2.getValue()){
                        return 1;
                    } else{
                        return m1.getKey().compareToIgnoreCase(m2.getKey());
                    }
                })
                .forEach(m -> System.out.printf("%s: %s\n", m.getKey(), m.getValue()));

        junkMaterials.entrySet().stream()
                .sorted((m1, m2) -> m1.getKey().compareTo(m2.getKey()))
                .forEach(m -> System.out.printf("%s: %s\n", m.getKey(), m.getValue()));
    }
}
